### ISSUE TITLE
Regression with Marshal.load(a Hash)

### ISSUE THREAD

**Issue by larskanis** [jruby/jruby#5523]
### Environment

```sh
$ ruby -v
jruby 9.2.5.0 (2.5.0) 2018-12-06 6d5a228 OpenJDK 64-Bit Server VM 10.0.2+13-Ubuntu-1ubuntu0.18.04.2 on 10.0.2+13-Ubuntu-1ubuntu0.18.04.2 +jit [linux-x86_64]

$ uname -a
Linux netzbuch 4.15.0-36-generic #39-Ubuntu SMP Mon Sep 24 16:19:09 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

### Expected Behavior

This is what MRI 2.5.3 and JRuby <= 9.2.4.0 computes:
```ruby
h={:€a=>nil, :€c=>nil, :€h=>nil}
Marshal.load(Marshal.dump(h)) #  => {:€a=>nil, :€c=>nil, :€h=>nil}
```

### Actual Behavior

But with JRuby-9.2.5.0 wired things are happening:
```ruby
h={:€a=>nil, :€c=>nil, :€h=>nil}
Marshal.load(Marshal.dump(h)) #  => {:€a=>nil, :€c=>true, nil=>:€h}
```

This regression was raised by the [Eventbox](https://github.com/larskanis/eventbox) test suite.

**Comment by headius** [comment]
Nice find! Maybe we should be running Eventbox's suite in our CI 😁

**Comment by headius** [comment]
Ok, I see the problem. It looks like a flaw in how we register a symbol and its encoding in the marshaled data. The indices of the links are getting off, which causes it to ready too little data for `€c`, so the next value in the stream is the "T" for UTF-8 encoding, and then the next key appears to be the nil value that `€c` should have been assigned.

**Comment by headius** [comment]
I believe I've fixed the issue in #5528, though there's a bit of a caveat listed there relating to visibility of partially-unmarshaled symbols.

The problem was largely as I described it above; however, the bug was in unmarshaling rather than marshaling. On the dump side, we register the symbol first and then its encoding, but on the unmarshal side we accidentally registered the encoding first. This led to it loading the wrong content for subsequent encodings since they all used the same encoding key of `:E`. The links for any encodings past the first would point at the first symbol, see it as not an encoding, and then not try to unmarshal the actual encoding. This shifted all values in the marshal stream around, so the keys and values were thrown off.

**Comment by larskanis** [comment]
Thanks @headius for fixing this so fast!
> Maybe we should be running Eventbox's suite in our CI :grin:

It doesn't pass on JRuby currently due to issues in `Thread.raise`/`Thread.handle_interrupt` which Eventbox makes use of. I didn't dig deeper into it to build a minimal example, since the related MRI tests are [still disabled in JRuby](https://github.com/jruby/jruby/blob/66d2905ae233a39e36fcbe3ade6382c2892ade8e/test/mri/excludes/TestThread.rb). It would be nice if they could be addressed sometime.
